[WIP] Core/Misc Support for Account Wide Mounts

Boomper opened the pull request:
Everything from TODO list seems to be finished (except playercondition.db2, no idea how that works, I'll leave it to someone else) so feel free to test it out and leave feedback.

Edit: Currently I'm stuck with 6.2 client unable to downgrade it/use previous versions (#134 error) so waiting till the core get's updated.

Boomper committed:
Core/Misc Support for Account Wide Mounts

 - it's still work in progress, I'm just tired from this weird weather so I'd like some feedback before I mess something up...
 - the packet size is based on https://gist.github.com/Boomper/4434a33ac3c0f10dd1ee packet length was  476 and if you fill it with numbers (1 + 4 + 4 + 452 + 14 + 1) it equals 476 aswell... so it propably makes sense...
 - TODO:
  - Evil SQL Query to transfer old mounts from character_spell into account_mounts (I'll propably have to export the Mount.db2 for spellIds (just thinking out loud)
  - add support for PlayerCondition.db2
  - add support for db table disables
  - add support for mounts that change when you change your faction (propably table in world db player_factionchange_spells)

Vincent-Michael commented on src/server/game/Server/Packets/MiscPackets.h:
```diff
@@ -663,6 +663,29 @@ namespace WorldPackets
+            void InitializeMounts(std::unordered_map<uint32, bool> mounts, bool fullUpdate);
+            WorldPacket const* Write() override;
+
+            bool IsFullUpdate;
```
bool IsFullUpdate = false;

VirusEntertainment commented:
gj

Shauren commented:
You are executing database queries directly in packet handlers (MountSetFavorite) - this is unacceptable as it opens a DoS attack possibility. I suggest you take a look at how everything else is saved for players and add it there (Player::SaveToDB)

Boomper committed:
Core/Misc Account Wide Mounts

 - actual "save system"
 - account_mounts table is now moved into LoginDatabase (auth)

Boomper committed:
DB/Misc TYPO

 - forgot to move the sql file...

jackpoz commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -3420,6 +3420,9 @@ bool Player::AddSpell(uint32 spellId, bool active, bool learning, bool dependent
         return false;
     }
 
+    if (MountEntry const* mountEntry = sDB2Manager.GetMount(spellId))
```
mountEntry variable looks unused

jackpoz commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -26750,3 +26762,73 @@ void Player::SendSpellCategoryCooldowns()
+
+bool Player::AddMount(uint32 spellId, bool isFavorite /*= false*/, bool isNew /*= false*/)
+{
+    SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(spellId);
```
few unused variables here too. Is this spellid check needed or is the GetMount() check below enough ?

Boomper commented:
MountEntry is there for "Todo: add support for PlayerCondition.db2" because it contains value "PlayerConditionId" which is for mounts that you can use only after you complete quest/achievement/something...

Currently the loading part isn't working, must've messed something up when I've moved the table into LoginDatabase...

Boomper commented:
Just finished writing the "Evil SQL Query to transfer old mounts from character_spell into account_mounts"...
https://gist.github.com/Boomper/ea3f851f66ddf0f8f1eb
Since it's "touching" auth, characters and world database (multidb queries) should it still be in sql/updates/auth or is there a special place for it?

I'll be updating the pull request in few hours, just need to polish some stuff (the loading should be fixed by then)...

Aokromes commented:
The query must be on the final destination table.

Boomper committed:
DB/Misc Updated query for account wide mount system

 - now also merges all spells from character_spell into account_mounts
 - not going to commit other things tonight, my computer is being mean by refusing to compile core under 30 minutes so rest will be tomorrow (today 10.6.2015) night-ish

Boomper commented on sql/updates/auth/2015_MM_DD_NN_auth.sql:
```diff
@@ -0,0 +1,532 @@
+JOIN `characters`.`characters` `chars` ON `chars`.`guid` = `characters`.`character_spell`.`guid`
+WHERE `characters`.`character_spell`.`spell` IN (SELECT `spell` FROM `temp_mounts`);
+
+DELETE FROM `characters`.`character_spell` WHERE `spell` IN (SELECT `spell` FROM `temp_mounts`);
```
@Aokromes I was writing about this. The reason why I don't want it in a seperate sql file is that it deletes the now useless mounts from character_spell and uses `temp_mounts` table which is dropped afterwards... -> it wouldn't do anything because the table temp_mounts wouldn't exist.
Yes I could just do "WHERE `spell` IN (1, 2,...)" but that would open up the possiblity that someone somewhere could execute the "DELETE FROM..." query before this one which would make him loose all playermounts.

Aokromes replied:
Maybe the way to do is a secondary commit after few days on that way no one fucks the database xd

Warpten replied:
the problem with this sql file is that you directly reference world, chars, characters (dafuq?) databases in this file, while there is no guarantee for anyone that all databases use those names

Shauren replied:
Or even that these databases are on the same server

Boomper replied:
From now on I shall reference them through time-space rift..., removing this from the pull request in next update, thanks for those heartwarming lines with so much potential to help me with this.

Boomper committed:
Core/Misc Update of AccountWide MountSystem

 - loading from auth DB now works
  - but I need to look into the asynch loading for smoother experience (can't quite figure out how that works for login db, any pointers would be appreciated (I don't mean pointers like 0x2E314A81))
 - added basic handling for faction mounts (through world db table playerfactionchange_spells), players now learn both version (alliance and horde) but only the version for their faction will appear in their mount collection (there are exceptions because some faction based mounts aren't included in that table
 - also added m_mountCount variable into Player class for future handling of mount count achievements (+ for size of vectors with sweet packet data (this text doesn't have to make sense, I'm tired))

Boomper committed:
TYPO

 - whooooops... hopefuly no one notices...

Boomper commented:
Wrongly merged something, I'll fix it during sunlight...

zwj-2014 commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -26750,3 +26764,96 @@ void Player::SendSpellCategoryCooldowns()
+    if (!sSpellMgr->GetSpellInfo(spellId))
+        return false;
+
+    if (!sDB2Manager.GetMount(spellId))
```
check again?

zwj-2014 commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -3420,6 +3422,9 @@ bool Player::AddSpell(uint32 spellId, bool active, bool learning, bool dependent
     }
 
+    if (sDB2Manager.GetMount(spellId))
+        return AddMount(spellId, false, true);
```
here

zwj-2014 commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -3420,6 +3422,9 @@ bool Player::AddSpell(uint32 spellId, bool active, bool learning, bool dependent
         return false;
     }
 
+    if (sDB2Manager.GetMount(spellId))
```
here

Boomper replied:
If I remove the one in AddSpell all spells will go into AddMount which just looks bad.
If I remove the one in AddMount I'm risking that a spell which used to be mount but now isn't is loaded as mount (and it's general prevention for data corruption inside that table).

jackpoz commented:
> Just finished writing the "Evil SQL Query to transfer old mounts from character_spell into account_mounts"...
> https://gist.github.com/Boomper/ea3f851f66ddf0f8f1eb

lines like
```sql
DROP TABLE IF EXISTS `auth`.`account_mounts`;
```
don't assure in ANY way that the statement is executed on the correct database and should NEVER be written in that way (speaking of personal experience, in my case "auth" is 335 auth database and "auth6x" is 6.x auth database, so your sql will execute statements on the 335 database instead of the 6.x one)

This is just the issue and it doesn't include the solution. https://github.com/TrinityCore/TrinityCore/commit/919ab9157e7458257e300546e05312049b0872cd is what I did last time I encountered this issue, up to you (and other devs) about what to do.

Take a look how the solution plays with TC Database Auto Update too.

Boomper committed:
Core/Misc TYPO

 - wrongly merged files from yesterday (did some cleanup in my local branches, from now on there shouldn't be issues like this)

Boomper committed:
DB/Misc Remove of conversion part of query for account_mounts table

 - it's going to be replaced with space-time rift which will convert all the old mounts in character_spell into those new ones (long story short: I'm not going to do anything with this until someone suggests better way)

Boomper commented:
Fixed the compilation errors from yesterday (problems with my local branches), removed the conversion part from SQL file and updated the description.

Boomper committed:
Core/Player Async Loading of Account-Wide Mounts

 - added new sqlqueryholder for other account wide things like achievements, battle pets, quets,...
 - accepting suggestions for better name for "collectionsHolder"

Boomper committed:
Core/Player Convesion of old (character_spell) mounts into new mounts (account_mounts)

 - old mounts are now added into new ones when player enters the world
 - no idea how that async query converted itself into sync query... it should be async... weird

Boomper commented:
The merging of old mounts is now handled by core (when player enters the world), I think that's all I was planning on doing so can someone test this if they find something wrong with it so we can finish this?

Frytiks commented:
Nice work, keep it :)

Boomper commented:
So I did some more research into the account wide mounts system and it seems like that @Exodius was right. I'm currenlty trying to write a list for sql table for mounts that should be account-wide/account-wide in spellbook/and only character specific...

Shauren commented:
@Boomper don't you dare do that. The client is able to know this by itself (see ingame mount tooltips) - fd786c83066a7ff165741915687c7c775ad54a16

Boomper commented:
@Shauren Completely missed that one, thanks :+1:
I did found that there are few mount "pairs" like:
http://www.wowhead.com/item=44226 / http://www.wowhead.com/item=44225 (not sure why) http://www.wowhead.com/item=67101 / http://www.wowhead.com/item=62298 (unlocked by completing achievement)
That work in a way that If you learn one of them you also learn the second one and you're only able to see one of them in your mounts (depending on your faction).
So what I'm currently thinking is that I should still make a table to list these mounts, and maybe in same table I should keep track of faction for which the mount works (doesn't seem to be stored in DBC) and temporarily raceMask/classMask till playercondition.db2 is implemented.

Shauren commented:
Just a heads up - I am working on implementing something nice for loading account wide data (it will not be done like you have it)

Boomper commented:
https://gist.github.com/Boomper/e6a2c9abc45512ab5364
That's the table that I'm proposing (accepting suggestions for better table/variable names), how it should work:
 - spell (spell id of the mount)
 - faction (0 - factionism free; 1 - alliance only; 2 - horde only) - if the mount is horde only on alliance character, he/she will see it in his collections but it will be red (unusable)
 - other_spell (oposite version) - those are special cases when the player gets both from earning one of them but can only see one of them (his faction version) in his collection
 - raceMask (1 Human | 2 Orc |...) - these mounts will only appear in collection if player fits into raceMask
 - classMask (same as raceMask)
I should have the table implemented mid-week if I get enough free time.
P.S.: Too tired, going to cry... so much gold/time wasted on testing.

Shauren commented:
> - other_spell (oposite version) - those are special cases when the player gets both from earning one of them but can only see one of them (his faction version) in his collection

Can you give an example of this?

Boomper commented:
http://www.wowhead.com/spell=60424 Mekgineer's Chopper / http://www.wowhead.com/spell=55531 Mechano-Hog
It's mostly mounts that are rewarded from completing faction specific reputations/achievements/quests or mounts that are faction specific and award achievement for getting them (mekgineer's chopper/mechano-hog - achievement http://www.wowhead.com/achievement=2097)

Boomper committed:
Core/Something Support for mount definitions

 - just adding this so I can update my repo for 6.2 so I can finally test this out

Boomper committed:
Merge to f45ae7af431402e6d9e4aff488a4cc64b9dfffd3

 - TODO: Research how to squish commits... for reasons... it looks cooler

Aokromes commented:
All checks have failed
1 failing check
Details continuous-integration/travis-ci/pr — The Travis CI build failed
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

Boomper commented:
I've been busy with exams and studying for them, no idea when I'll have time to finish this.

tkrokli commented:
To fix the conflicts, you will usually only need to update/rebase your branch so it is even with 6.x
(just make sure that you got a backup copy of your files, in case anything goes wrong during rebase).

MitchesD commented:
I am closing this PR:
1) It's outdated
2) I am working on it ;)
3) Author looks like gone

Boomper commented:
Go for it, feel free to reuse whatever you want from it, if there's something worth a while.
I don't have time to work on this anymore.
